sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/ShardReadersPool.java

```diff
@@ -103,11 +151,15 @@ private void readLoop(ShardRecordsIterator shardRecordsIterator) {
 
   CustomOptional<KinesisRecord> nextRecord() {
     try {
-      KinesisRecord record = recordsQueue.poll(1, TimeUnit.SECONDS);
+      if (recordsQueue.get() == null) {
```

Comment:
This can only happen if start() hasn't been called, right? I think in this case we should throw an error.

Reply:
Or when shardIteratorsMap was empty in the start method. As I said in some previous comment, I didn’t find a way to tell Beam that the source that is using this pool will not produce records anymore.

Reply:
Can we change the start method to always create a queue? If the map is initially empty, just create a queue with a small capacity. That would make the invariants of the code more clear.